### PR TITLE
fix/improve tests for build

### PR DESCRIPTION
**Pull request by J-N-K**
Addresses some parts of #5775 

Signed-off-by: Jan N. Klug <jan.n.klug@rub.de>

### CODE REVIEW

**Comment by martinvw** [comment]
Could you enable the tests in travis, it would show the progres here :+

**Comment by martinvw** [comment]
@J-N-K feel free to check the boxes in the issue when things are fixed

**Comment by J-N-K** [comment]
MQTT is impossible to fix without #5503. We need the embedded broker, that needs moquette and that is not an OSGi bundle and needs to be wrapped. AFAIK bnd can't do that on-the-fly while resolving bndrun-files.

**Comment by martinvw** [comment]
> MQTT is impossible to fix without #5503. We need the embedded broker, that needs moquette and that is not an OSGi bundle and needs to be wrapped. AFAIK bnd can't do that on-the-fly while resolving bndrun-files.

Time to rebase ;-)

**Comment by martinvw** [comment]
Serious flaw in the new way of running the build, the status is not recognized correctly.

The build is passed but it did fail...

**Comment by martinvw** [comment]
> Serious flaw in the new way of running the build, the status is not recognized correctly.
> 
> The build is passed but it did fail...

And solved ;-)

File: buildci.sh [hunk 1]
```diff
@@ -35,6 +35,7 @@ if [ ! -z "$CHANGED_DIR" ] && [ -e "bundles/$CHANGED_DIR" ]; then
     echo "New addon pull request: Building $CHANGED_DIR"
     echo "MAVEN_OPTS='-Xms1g -Xmx2g -Dorg.slf4j.simpleLogger.log.org.openhab.tools.analysis.report.ReportUtility=DEBUG -Dorg.slf4j.simpleLogger.defaultLogLevel=WARN'" > ~/.mavenrc
     cd "bundles/$CHANGED_DIR"
+    mvn clean install -B 2>&1 |
```
Comment (davidgraeff) [inline buildci.sh]: This looks wrong to me

Comment (martinvw) [reply buildci.sh]: I hope this it the reason that the script is failing to fail to build, because it looked so promising...

**Comment by J-N-K** [comment]
There is still something wrong with Travis. It only builds the last commit.

**Comment by martinvw** [comment]
> There is still something wrong with Travis. It only builds the last commit.

No this is actually a feature, it only checks for changed bundles and if there is exactly one changed bundle it will build that one and its it-tests

**Comment by J-N-K** [comment]
But so we can only see the last changed bundles. Look at the build, it contains only the embedded broker. But in all commitsl I changed other bundles, too.

File: bundles/org.openhab.io.mqttembeddedbroker/src/test/java/org/openhab/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerServiceTest.java [hunk 2]
```diff
@@ -175,9 +186,9 @@ public void testPersistence() throws InterruptedException, IOException, Executio
 
         // Stop server -> close persistence storage and sync it to disk
         subject.deactivate();
-
-        File jksFile = new File(subject.getPersistenceFilename());
         assertTrue(jksFile.exists());
+        // this is needed to ensure the file is correctly written
+        waitForAssert(() -> assertEquals(12288, jksFile.length()));
```
Comment (martinvw) [inline bundles/org.openhab.io.mqttembeddedbroker/src/test/java/org/openhab/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerServiceTest.java]: isn't this platform dependent any other way to check?

**Comment by martinvw** [comment]
> But so we can only see the last changed bundles. Look at the build, it contains only the embedded broker. But in all commitsl I changed other bundles, too.

It literally only checks the folder `bundles` now, so it ignores the fact that you committed in the folder `ittest`, I will clone your branch and do some more tuning :-)

File: bundles/org.openhab.io.mqttembeddedbroker/src/test/java/org/openhab/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerServiceTest.java (hunk 2 above)
Comment (J-N-K) [reply bundles/org.openhab.io.mqttembeddedbroker/src/test/java/org/openhab/io/mqttembeddedbroker/internal/MqttEmbeddedBrokerServiceTest.java]: Not sure. It seems to work on windows and linux. In fact this is only an ugly workaround. There is no way (at least none that I know of) to check whether the broker completed its shutdown.  Before that is the case, the file is not completely written and the test will fail.

**Comment by davidgraeff** [comment]
I have to admit, I was lazy and only considered the bundles directory because itests weren't working at that time. But what should have happened is that if something outside of bundles (eg itests) gets changed, a full build should be started.

**Comment by martinvw** [comment]
> I have to admit, I was lazy and only considered the bundles directory because itests weren't working at that time. But what should have happened is that if something outside of bundles (eg itests) gets changed, a full build should be started.

I came up with another idea, are you okay with it?

https://github.com/openhab/openhab2-addons/pull/5797

**Comment by J-N-K** [comment]
That's all I can do for now. All tests now start, but some Home, HomeAssistant and HueEmulation still fail. HueEmulation needs the UpnpServiceConfiguration, I have no idea how to provide that. Not sure about the MQTT tests, I would like to see them on Travis, but this is currently not executed.

**Comment by davidgraeff** [comment]
Thanks @J-N-K

File: itests/org.openhab.binding.mqtt.homeassistant.tests/pom.xml [hunk 3]
```diff
@@ -13,6 +13,11 @@
 
   <name>openHAB Add-ons :: Integration Tests :: MQTT HomeAssistant Tests</name>
 
+       <properties>
```
Comment (martinvw) [inline itests/org.openhab.binding.mqtt.homeassistant.tests/pom.xml]: Small misalignment

File: itests/org.openhab.binding.mqtt.homeassistant.tests/pom.xml [hunk 4]
```diff
@@ -29,11 +34,72 @@
...
-      <groupId>org.openhab.addons.bundles</groupId>
-      <artifactId>org.openhab.io.mqttembeddedbroker</artifactId>
-      <version>${project.version}</version>
+        <dependency>
+            <groupId>org.openhab.addons.bundles</groupId>
+            <artifactId>org.openhab.io.mqttembeddedbroker</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+      <groupId>org.openhab.osgiify</groupId>
```
Comment (martinvw) [inline itests/org.openhab.binding.mqtt.homeassistant.tests/pom.xml]: Looks like a mix of tabs and spaces

File: itests/org.openhab.binding.mqtt.homie.tests/pom.xml [hunk 5]
```diff
@@ -1,39 +1,104 @@
...
+			<artifactId>org.openhab.binding.mqtt.homie</artifactId>
+			<version>${project.version}</version>
+		</dependency>
+		<dependency>
+            <groupId>org.openhab.addons.bundles</groupId>
+            <artifactId>org.openhab.io.mqttembeddedbroker</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+      <groupId>org.openhab.osgiify</groupId>
```
Comment (martinvw) [inline itests/org.openhab.binding.mqtt.homie.tests/pom.xml]: Mix of whitespace

**Comment by martinvw** [comment]
> That's all I can do for now.

I think we can merge this if the whitespace is fine, maybe you can execute the format pom command which @wborn once used on the repo:

```
mvn \
  com.github.ekryd.sortpom:sortpom-maven-plugin:sort \
  -Dsort.createBackupFile=false \
  -Dsort.keepBlankLines=true \
  -Dsort.predefinedSortOrder="recommended_2008_06"
```

>  I would like to see them on Travis, but this is currently not executed.

With a small hack (changing the buildci.sh) or merging my open PR that should be possible, the small hack is trivial, just add

```
CHANGED_DIR=

```

before

```
if [[ ! -z "$CHANGED_DIR" ]] && [[ -e "bundles/$CHANGED_DIR" ]]; then

File: features/openhab-addons/pom.xml [hunk 6]
```diff
@@ -35,25 +35,25 @@
...
-                    <exclude name="**/org.openhab.binding.bluetooth*/**/feature.xml"/>
-                    <exclude name="**/org.openhab.binding.mqtt*/**/feature.xml"/>
+                    <include name="*/src/main/feature/feature.xml"></include>
+                    <exclude name="**/org.openhab.binding.bluetooth*/**/feature.xml"></exclude>
+                    <exclude name="**/org.openhab.binding.mqtt*/**/feature.xml"></exclude>
                   </fileset>
                   <filterchain>
                     <linecontainsRegExp>
-                      <regexp pattern="(feature>)|(feature\s)|(bundle>)|(bundle\s)"/>
+                      <regexp pattern="(feature&gt;)|(feature\s)|(bundle&gt;)|(bundle\s)"></regexp>
```
Comment (martinvw) [inline features/openhab-addons/pom.xml]: Did you check whether this has no unexpected side effects?

File: buildci.sh [hunk 7]
```diff
@@ -73,7 +73,7 @@ if [[ ! -z "$CHANGED_DIR" ]] && [[ -e "bundles/$CHANGED_DIR" ]]; then
 else
     echo "Build all"
     echo "MAVEN_OPTS='-Xms1g -Xmx2g -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn'" > ~/.mavenrc
-    mvnp clean install -B -DskipChecks=true -DskipTests=true
+    mvnp clean install -B -DskipChecks=true
```
Comment (martinvw) [inline buildci.sh]: I fear we are not yet ready for this :(

File: features/openhab-addons/pom.xml (hunk 6 above)
Comment (J-N-K) [reply features/openhab-addons/pom.xml]: No. Didn’t see that. Came out of the automatic formatter. Will check the resulting feature-file.

File: buildci.sh (hunk 7 above)
Comment (davidgraeff) [reply buildci.sh]: And it takes to long to build with tests doesn't it?

Comment (martinvw) [reply buildci.sh]: > And it takes to long to build with tests doesn't it?

No I believe it justs fits but when using both checks and tests it becomes critical

File: features/openhab-addons/pom.xml (hunk 6 above)
Comment (J-N-K) [reply features/openhab-addons/pom.xml]: md5-hash is the same for both. so we can keep the new one

File: buildci.sh (hunk 7 above)
Comment (J-N-K) [reply buildci.sh]: done

**Comment by martinvw** [comment]
Many thanks @J-N-K

**Pull request merged**